module/zfs/arc.c

```diff
@@ -7743,12 +7883,22 @@ arc_init(void)
 
 	buf_init();
 
+	if (zfs_arc_evict_threads == 0)
+		zfs_arc_evict_threads_live = MIN(MAX(max_ncpus > 6 ? 2 : 1,
+		    arc_ilog2(max_ncpus) + (max_ncpus >> 6)), 16);
+	else
+		zfs_arc_evict_threads_live = zfs_arc_evict_threads;
+
 	list_create(&arc_prune_list, sizeof (arc_prune_t),
 	    offsetof(arc_prune_t, p_node));
 	mutex_init(&arc_prune_mtx, NULL, MUTEX_DEFAULT, NULL);
 
 	arc_prune_taskq = taskq_create("arc_prune", zfs_arc_prune_task_threads,
 	    defclsyspri, 100, INT_MAX, TASKQ_PREPOPULATE | TASKQ_DYNAMIC);
+	arc_evict_taskq = taskq_create("arc_evict",
+	    MIN(zfs_arc_evict_threads_live, max_ncpus), defclsyspri,
+	    MIN(zfs_arc_evict_threads_live, max_ncpus), max_ncpus,
```

Comment:
Have you got anything in particular on your mind? 

What I see here is that perhaps it is not right to cap the thread count at the CPU count, which might be against what the user when setting the tunable manually.

Reply:
Yea. When we calculate `zfs_arc_evict_threads_live` above -- it should be sane already.  If it is submitted by user -- should we really prevent foot-shooting?  And in any case I would move it out of `taskq_create()` call, limiting the `zfs_arc_evict_threads_live` value, or the `zfs_arc_evict_threads` if you remove duplication as I have proposed below.

Reply:
I'd agree that the foot-shooting prevention seems unnecessary. 

From what I understand:

- `nthreads` controls the number of threads in the queue, so we can just set it to `zfs_arc_evict_threads_live`
- `minalloc` controls the minimal size of the task entries cache. It guess it makes sense to set it to at least `zfs_arc_evict_threads_live`.
- `maxalloc` controls the maximum size of the task entries cache. I've grepped the tree for other examples of `taskq_create()` invocations and it seems like `INT_MAX` is a fairly popular value here. Do you have any thoughts about it?

The resulting code would look like this:

```c
	arc_evict_taskq = taskq_create("arc_evict",
	    zfs_arc_evict_threads_live, defclsyspri,
	    zfs_arc_evict_threads_live, max_ncpus, INT_MAX,
	    TASKQ_PREPOPULATE);
```

Reply:
About `nthread` I agree -- `zfs_arc_evict_threads_live`. `minalloc` generally you may set to a number of tasks you expect to be scheduled same time often enough, that may avoid some memory allocations, especially when we may be low on memory. `maxalloc` I would set to INT_MAX unless you want the code to actually limit that, which I don't think you do.  But since you are pre-allocating the taskq entries as part of `evict_arg_t` and using `taskq_dispatch_ent()` it is actually irrelevant.  You may just set the min and max to 0 and INT_MAX respectively.

Reply:
Thanks! I've applied the recommendations in the latest patch set I pushed.